### PR TITLE
Added CMake

### PR DESCRIPTION
**Pull request by bseto**
To run this.

1. Go to top level directory of TerminalFighter
2. create a new directory called 'build' `mkdir build`
3. run `cmake ..`
4. run `make` (you can use `make -j8` or whatever if you want parallel compiling - yes, it works)
5. binary will be located inside the build/ directory, inside src/ `TerminalFighter/build/src/`

In hindsight, I think travis will fail.

### CODE REVIEW

**Comment by ExtraGravity** [comment]
Should fix the travis build.

**Review by ExtraGravity — CHANGES_REQUESTED** [review]
- FindSDL2_ttf.cmake was taken from somewhere right?

- For travis, using the following wget, `https://l.messenger.com/l.php?u=https%3A%2F%2Fsourceforge.net%2Fprojects%2Fastyle%2Ffiles%2Fastyle%2Fastyle%25203.0%2Fastyle_3.0_linux.tar.gz%2Fdownload&h=ATMZImYXU7vmymhRfm0DIngfMBFkkyoei50lUJuQPU8HjwDV7g2yB3jauK7Jt2J7lp0oFImOy-n3Ozhxy9o77q8ayXz1r-eetuZLojLiybdCg4zgVGDp8n2XwJb4iHiBwAatRWGcONo`

- Lastly, I think the Gamestates should stay in it's own DIR, it'll need it's own CMakeLists.txt, and it's something we should probably learn to do now. Or we're not going to use multiple DIRS in the future just because we don't want to go through the hassle of learning how to set it up.

- README should be updated with build steps

File: FindSDL2_ttf.cmake [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        # path suffixes to search inside ENV{SDLDIR}
+        include/SDL2 include
+        PATHS ${SDL2_TTF_PATH}
+        )
+
+if (CMAKE_SIZEOF_VOID_P EQUAL 8)
+    set(VC_LIB_PATH_SUFFIX lib/x64)
+else ()
+    set(VC_LIB_PATH_SUFFIX lib/x86)
+endif ()
```
Comment (ExtraGravity) [inline FindSDL2_ttf.cmake:59]: This whole file you took from somewhere else right?

File: README.adoc [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+`Make -j8`
++
+. Copy the executable to the correct location
+`cp src/TerminalFighter ../`
+
+== How to Build Using `build-all.sh`
+
+. Run `build-all.sh` script
++
+`./build-all.sh`
```
Comment (ExtraGravity) [inline README.adoc:23]: You mean make-all? Well I think build-all is a better name now that we dont' use makefile directly. So renaming the mak-all script would be better.

File: src/gamestates/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+add_library(GamestateHandler GamestateHandler.cpp)
+add_library(MenuState MenuState.cpp)
+add_library(TestState TestState.cpp)
+
+target_link_libraries(TestState GameObjectMediator Universe CollisionDetector BasicEnemy GraphicsHandler MainCharacter MissileLauncher Events Delay Keyboard)
```
Comment (ExtraGravity) [inline src/gamestates/CMakeLists.txt]: Can we multiline this?

File: src/gamestates/I_Gamestate.h [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #pragma once
 
-#include <SDL.h>
+/* #include <SDL.h> */
```
Comment (ExtraGravity) [inline src/gamestates/I_Gamestate.h]: Delete?

File: src/gamestates/TestState.cpp [hunk 5]
```diff
@@ -2,12 +2,12 @@
 
 #include "TestState.h"
 
-#include "GameObjectMediator.h"
-#include "Universe.h"
-#include "CollisionDetector.h"
-#include "BasicEnemy.h"
-#include "GraphicsHandler.h"
-#include "MainCharacter.h"
+#include "src/GameObjectMediator.h"
```
Comment (ExtraGravity) [inline src/gamestates/TestState.cpp:5]: All this src/ is pretty ugly, can we get rid of it?

File: src/gamestates/TestState.h [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
 #include <SDL2/SDL.h>
 
-#include "I_Gamestate.h"
-#include "Delay.h"
-#include "GameConstants.h"
-#include "KeyboardListener.h"
-#include "Keyboard.h"
-#include "MissileLauncher.h"
-#include "Events.h"
+#include <src/gamestates/I_Gamestate.h>
```
Comment (ExtraGravity) [inline src/gamestates/TestState.h]: Use quotes instead of brackets.

File: test/TestAll.cpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#include <gtest/gtest.h>
+#include "CollisionDetectorTest.cpp"
+#include "GameObjectTest.cpp"
+#include "GameStateHandlertest.cpp"
+#include "HitboxTest.cpp"
+#include "KeyboardListenerTest.cpp"
+#include "MissileLauncherTest.cpp"
```
Comment (ExtraGravity) [inline test/TestAll.cpp:7]: You shouldn't ever need to include cpp files, that's a recipe for circular dependencies. Also, gtest is supposed ot have auotomatic test discovery that doesn't need this.

**Review by ExtraGravity — CHANGES_REQUESTED** [review]
Still need to test it on my machine.

File: FindSDL2_ttf.cmake (hunk 1 above)
Comment (bseto) [reply FindSDL2_ttf.cmake:59]: Yep

File: src/gamestates/CMakeLists.txt (hunk 3 above)
Comment (bseto) [reply src/gamestates/CMakeLists.txt]: Good idea

File: README.adoc (hunk 2 above)
Comment (bseto) [reply README.adoc:23]: Ok, I can do that

File: src/gamestates/TestState.cpp (hunk 5 above)
Comment (bseto) [reply src/gamestates/TestState.cpp:5]: Not sure :/, tried to look into it b4 but things I tried didn't work. Tried a big combination of things with the cmake and the file includes. Il give it one more shot though

Comment (ExtraGravity) [reply src/gamestates/TestState.cpp:5]: We can run the build within the src directory too, and the make all script will just cd into the src before running it's current stuff.

File: test/TestAll.cpp (hunk 7 above)
Comment (bseto) [reply test/TestAll.cpp:7]: W/o the includes, i get:
[==========] Running 0 tests from 0 test cases.
[==========] 0 tests from 0 test cases ran. (0 ms total)
[  PASSED  ] 0 tests.

Not sure why, but when we used a normal makefile, it'd discover the tests, but w/ CMake, it seems like i need the includes for these. 

Example where they do it: https://www.eriksmistad.no/getting-started-with-google-test-on-ubuntu/

File: src/gamestates/TestState.cpp (hunk 5 above)
Comment (bseto) [reply src/gamestates/TestState.cpp:5]: ? sorry i don't follow.

File: test/TestAll.cpp (hunk 7 above)
Comment (ExtraGravity) [reply test/TestAll.cpp:7]: I think there's a call in CMakeLists to add tests. https://cmake.org/cmake/help/v3.0/module/FindGTest.html I would try that, including cpp files can make some really funky errors :/

**Pull request merged**